### PR TITLE
Map.transform

### PR DESCRIPTION
**Pull request by boxxxie**
a combination of Map.remap_keys, and Map.transform_values.

### CODE REVIEW

File: test/kitchen_sink/map_test.exs [hunk 1]
```diff
@@ -160,4 +160,48 @@ defmodule KitchenSink.MapTest do
...
+    transformation_map = %{
+      age: {[:new_age], fn x -> x end},
+      gender: {:gender, &String.to_atom/1},
+      multiplier: {[:rate, :inner_rate],&String.to_float/1},
+      smoker?: {:smoker! ,&String.to_atom/1},
+      term: {&String.to_float/1},
+    }
+
+    actual = KMap.transform(input, transformation_map, prune: true)
+    assert actual == expected
```
Comment (kenips) [inline test/kitchen_sink/map_test.exs:191]: prune isn't actually tested

File: lib/kitchen_sink/map.ex [hunk 2]
```diff
@@ -240,4 +240,71 @@ defmodule KitchenSink.Map do
...
+  b: {[:new, :nested, :key, :name], b_transform_fun},
+  ...
+  }
+
+  output is a transformed Map, or function that takes a Map.  the output Map is made from appling each of the
+  transformers in the transformer-map to the corresponding keys and values in the Map, outputing a Map where each
+  key-value in the Map has been transformed. supplying `prune: true` prunes the map so only transformed values are
+  output.
+
+  fn(%{a:, b: ...}) -> %{a: a_transform_fun(a), b: b_transform_fun(b) ...}
```
Comment (kenips) [inline lib/kitchen_sink/map.ex]: Would be great if we add a small example and doctest it too.

**Review by kenips — CHANGES_REQUESTED** [review]
Tests are not passing

File: lib/kitchen_sink/map.ex [hunk 3]
```diff
@@ -240,4 +240,97 @@ defmodule KitchenSink.Map do
...
+       %{b1: %{b2: 0.99}}
+
+  output is a transformed Map.  the output Map is made from appling each of the
+  transformers in the transformer-map to the corresponding keys and values in the Map, outputing a Map where each
+  key-value in the Map has been transformed. supplying `prune: true` prunes the map so only transformed values are
+  output.
+  """
+  @lint {Credo.Check.Refactor.ABCSize, false}
+  @spec transform(Map.t, Map.t, Keyword.t) :: Map.t
+  def transform(map, key_value_transform_map, [prune: true] = _opts) do
```
Comment (olafura) [inline lib/kitchen_sink/map.ex:264]: So there is no options for non pruning?

File: lib/kitchen_sink/map.ex [hunk 4]
```diff
@@ -240,4 +240,97 @@ defmodule KitchenSink.Map do
...
+  key-value in the Map has been transformed. supplying `prune: true` prunes the map so only transformed values are
+  output.
+  """
+  @spec transform(Map.t, Map.t) :: Map.t
+  def transform(map, key_value_transform_map) do
+    keys_to_transform = Map.keys(key_value_transform_map)
+    map_without_transformed_keys = map |> Map.drop(keys_to_transform)
+
+    map
+    |> transform(key_value_transform_map, prune: true)
```
Comment (olafura) [inline lib/kitchen_sink/map.ex:315]: Where is this call, doesn't this just go into the same function looping?

File: lib/kitchen_sink/map.ex (hunk 3 above)
Comment (olafura) [inline lib/kitchen_sink/map.ex:264]: So we don't have a non prune version?

**Review by olafura — CHANGES_REQUESTED** [review]

File: test/kitchen_sink/map_test.exs (hunk 1 above)
Comment (boxxxie) [reply test/kitchen_sink/map_test.exs:191]: it is now.

File: lib/kitchen_sink/map.ex (hunk 3 above)
Comment (boxxxie) [reply lib/kitchen_sink/map.ex:264]: the options are yes or no. what other options do you think should exist?

Comment (boxxxie) [reply lib/kitchen_sink/map.ex:264]: we do have it. it's &transform/2

File: lib/kitchen_sink/map.ex (hunk 4 above)
Comment (boxxxie) [reply lib/kitchen_sink/map.ex:315]: test's wouldn't work in that case... would they?

File: lib/kitchen_sink/map.ex (hunk 3 above)
Comment (olafura) [reply lib/kitchen_sink/map.ex:264]: Which calls prune

**Review by olafura — APPROVED** [review]

**Review by kenips — APPROVED** [review]

**Pull request merged**